### PR TITLE
[ipcamera] Add Reolink API support

### PR DESCRIPTION
**Pull request by Skinah**
These changes add Reolink support which a lot of users have requested and been testing and giving feedback already in the forum here:

https://community.openhab.org/t/adding-reolink-api-to-the-ipcamera-binding-beta-testers-needed/142202 

This pull request will automatically be built and available under the following links if anyone wants to test:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.ipcamera/4.0.0-SNAPSHOT/org.openhab.binding.ipcamera-4.0.0-SNAPSHOT.jar

Signed-off-by: Matthew Skinner <matt@pcmus.com>

### CODE REVIEW

File: bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifConnection.java [hunk 1]
```diff
@@ -601,98 +620,134 @@ public void gotoPreset(int index) {
...
-            case "VideoSource/ImageTooBlurry/ImagingService":
-            case "VideoSource/ImageTooBlurry/RecordingService":
-                if ("true".equals(dataValue)) {
-                    ipCameraHandler.changeAlarmState(CHANNEL_TOO_BLURRY_ALARM, OnOffType.ON);
-                } else if ("false".equals(dataValue)) {
-                    ipCameraHandler.changeAlarmState(CHANNEL_TOO_BLURRY_ALARM, OnOffType.OFF);
-                }
-                break;
-            default:
+            switch (topic) {
```
[the remark this answers is not in the document]
Comment (Skinah) [reply bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/onvif/OnvifConnection.java:627]: I do not remember my reasoning, but I agree your suggestion is nicer to read. Applied suggestion.

**Comment by Skinah** [comment]
@jlaur Thank you for your suggestions which should be all addressed now.... I do not understand your hard coding comment above, perhaps because I am self taught and still learning basic coding concepts. The URL's are defined in an API and the binding is implementing the API as documented. If you could elaborate on what the potential issue is or the better way? Do you mean using constants to define the urls in the one place/file, in case the API changes in the future to reduce the workload changing many lines of code?

File: bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/handler/IpCameraHandler.java [hunk 2]
```diff
@@ -1710,6 +1808,11 @@ private void offline() {
             localFuture.cancel(true);
             localFuture = null;
         }
+        localFuture = authenticationJob;
+        if (localFuture != null) {
+            localFuture.cancel(true);
+            localFuture = null;
```
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/handler/IpCameraHandler.java]: Just checking, you don't want to fix the same bug above and below your change?

File: bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/ReolinkHandler.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                            "[{ \"cmd\":\"GetIrLights\"}]");
+                    break;
+            }
+            return;
+        } // end of "REFRESH"
+        switch (channelUID.getId()) {
+            case CHANNEL_ACTIVATE_ALARM_OUTPUT: // cameras built in siren
+                if (OnOffType.ON.equals(command)) {
+                    ipCameraHandler.sendHttpPOST("/api.cgi?cmd=AudioAlarmPlay" + ipCameraHandler.reolinkAuth,
+                            "[{\"cmd\": \"AudioAlarmPlay\", \"param\": {\"alarm_mode\": \"manul\", \"manual_switch\": 1, \"channel\": "
```
Comment (jlaur) [inline bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/ReolinkHandler.java:195]: Just checking, is "manul" correct, or should it be "manual"?

**Comment by jlaur** [comment]
> I do not understand your hard coding comment above, perhaps because I am self taught and still learning basic coding concepts

What I'm referring to is code like this:
```java
ipCameraHandler.sendHttpPOST("/api.cgi?cmd=AudioAlarmPlay" + ipCameraHandler.reolinkAuth,
    "[{\"cmd\": \"AudioAlarmPlay\", \"param\": {\"alarm_mode\": \"manul\", \"manual_switch\": 1, \"channel\": "
    + ipCameraHandler.cameraConfig.getNvrChannel() + " }}]");
```

While it will probably work perfectly well, it's hard to read, and for the same reason trivial bugs (like syntax errors) could easily be introduced unnoticed. If this JSON would be serialized from a DTO, it would be easier to read and the risk of syntax errors would be eliminated.

File: bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/handler/IpCameraHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Skinah) [reply bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/handler/IpCameraHandler.java]: Thank you, I have fixed the other bugs that are the same.

File: bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/ReolinkHandler.java (hunk 3 above)
Comment (Skinah) [reply bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/ReolinkHandler.java:195]: It is correct, the Reolink API contains a number of spellings that look wrong.

**Comment by Skinah** [comment]
Thank you for the suggestions and I believe I have completed all requested changes.

Regarding your non requested suggestion of DTO, I did a quick look and do not wish to go that route, but someone in future can do it. Considering that I do not own one of these brands to do any testing with, I do not have the time when this basic method is going to suit 99% of people just fine and most of it is already tested by end users.

There are at least 2 users of Reolink cameras that can code in Java that are waiting for this to be merged so they can take over and move it forward so it is possible someone will make that change that is able to do tests on real hardware.

**Review by jlaur — APPROVED** [review]
LGTM

**Pull request merged**